File: src/python/ip/src/histo.cc
/**
 * @file python/ip/src/histo.cc
 * @date Mon Apr 18 16:08:34 2011 +0200
 * @author Francois Moulin <Francois.Moulin@idiap.ch>
 *
 * @brief Binds histogram to python
 *
 * Copyright (C) 2011 Idiap Reasearch Institute, Martigny, Switzerland
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "core/python/ndarray.h"
#include "ip/histo.h"

using namespace boost::python;
namespace ip = Torch::ip;
namespace tp = Torch::python;
namespace ca = Torch::core::array;

template <typename T>
static object inner_histo1 (tp::const_ndarray input) {
  int size = Torch::ip::detail::getHistoSize<T>();
  tp::ndarray out(ca::t_uint64, size);
  blitz::Array<uint64_t,1> out_ = out.bz<uint64_t,1>();
  ip::histogram(input.bz<T,2>(), out_, false);
  return out.self();
}

static object histo1 (tp::const_ndarray input) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_uint8: return inner_histo1<uint8_t>(input);
    case ca::t_uint16: return inner_histo1<uint16_t>(input);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

template <typename T>
static void inner_histo2 (tp::const_ndarray input, tp::ndarray output,
    bool accumulate) {
  blitz::Array<uint64_t,1> out_ = output.bz<uint64_t,1>();
  ip::histogram(input.bz<T,2>(), out_, accumulate);
}

static void histo2 (tp::const_ndarray input, tp::ndarray output,
    bool accumulate=false) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_uint8: return inner_histo2<uint8_t>(input, output, accumulate);
    case ca::t_uint16: return inner_histo2<uint16_t>(input, output, accumulate);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

BOOST_PYTHON_FUNCTION_OVERLOADS(histo2_overloads, histo2, 2, 3)

template <typename T>
static void inner_histo3 (tp::const_ndarray input, tp::ndarray output,
    object max, bool accumulate) {
  blitz::Array<uint64_t,1> out_ = output.bz<uint64_t,1>();
  T tmax = extract<T>(max);
  ip::histogram(input.bz<T,2>(), out_, (T)0, tmax, (uint32_t)(tmax+1), accumulate);
}

static void histo3 (tp::const_ndarray input, tp::ndarray output, object max,
    bool accumulate=false) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_int8: 
      return inner_histo3<int8_t>(input, output, max, accumulate);
    case ca::t_int16: 
      return inner_histo3<int16_t>(input, output, max, accumulate);
    case ca::t_int32: 
      return inner_histo3<int32_t>(input, output, max, accumulate);
    case ca::t_int64: 
      return inner_histo3<int64_t>(input, output, max, accumulate);
    case ca::t_uint8: 
      return inner_histo3<uint8_t>(input, output, max, accumulate);
    case ca::t_uint16: 
      return inner_histo3<uint16_t>(input, output, max, accumulate);
    case ca::t_uint32: 
      return inner_histo3<uint32_t>(input, output, max, accumulate);
    case ca::t_uint64: 
      return inner_histo3<uint64_t>(input, output, max, accumulate);
    case ca::t_float32: 
      return inner_histo3<float>(input, output, max, accumulate);
    case ca::t_float64: 
      return inner_histo3<double>(input, output, max, accumulate);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

BOOST_PYTHON_FUNCTION_OVERLOADS(histo3_overloads, histo3, 3, 4)

template <typename T>
static void inner_histo4 (tp::const_ndarray input, tp::ndarray output,
    object min, object max, bool accumulate) {
  blitz::Array<uint64_t,1> out_ = output.bz<uint64_t,1>();
  T tmin = extract<T>(min);
  T tmax = extract<T>(max);
  ip::histogram(input.bz<T,2>(), out_, tmin, tmax, (uint32_t)(tmax-tmin+1), accumulate);
}

static void histo4 (tp::const_ndarray input, tp::ndarray output, 
    object min, object max, bool accumulate=false) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_int8: 
      return inner_histo4<int8_t>(input, output, min, max, accumulate);
    case ca::t_int16: 
      return inner_histo4<int16_t>(input, output, min, max, accumulate);
    case ca::t_int32: 
      return inner_histo4<int32_t>(input, output, min, max, accumulate);
    case ca::t_int64: 
      return inner_histo4<int64_t>(input, output, min, max, accumulate);
    case ca::t_uint8: 
      return inner_histo4<uint8_t>(input, output, min, max, accumulate);
    case ca::t_uint16: 
      return inner_histo4<uint16_t>(input, output, min, max, accumulate);
    case ca::t_uint32: 
      return inner_histo4<uint32_t>(input, output, min, max, accumulate);
    case ca::t_uint64: 
      return inner_histo4<uint64_t>(input, output, min, max, accumulate);
    case ca::t_float32: 
      return inner_histo4<float>(input, output, min, max, accumulate);
    case ca::t_float64: 
      return inner_histo4<double>(input, output, min, max, accumulate);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

BOOST_PYTHON_FUNCTION_OVERLOADS(histo4_overloads, histo4, 4, 5)

template <typename T>
static void inner_histo5 (tp::const_ndarray input, tp::ndarray output,
    object min, object max, uint32_t nbins, bool accumulate) {
  blitz::Array<uint64_t,1> out_ = output.bz<uint64_t,1>();
  T tmin = extract<T>(min);
  T tmax = extract<T>(max);
  ip::histogram(input.bz<T,2>(), out_, tmin, tmax, nbins, accumulate);
}

static void histo5 (tp::const_ndarray input, tp::ndarray output, 
    object min, object max, uint32_t nbins, bool accumulate=false) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_int8: 
      return inner_histo5<int8_t>(input, output, min, max, nbins, accumulate);
    case ca::t_int16: 
      return inner_histo5<int16_t>(input, output, min, max, nbins, accumulate);
    case ca::t_int32: 
      return inner_histo5<int32_t>(input, output, min, max, nbins, accumulate);
    case ca::t_int64: 
      return inner_histo5<int64_t>(input, output, min, max, nbins, accumulate);
    case ca::t_uint8: 
      return inner_histo5<uint8_t>(input, output, min, max, nbins, accumulate);
    case ca::t_uint16: 
      return inner_histo5<uint16_t>(input, output, min, max, nbins, accumulate);
    case ca::t_uint32: 
      return inner_histo5<uint32_t>(input, output, min, max, nbins, accumulate);
    case ca::t_uint64: 
      return inner_histo5<uint64_t>(input, output, min, max, nbins, accumulate);
    case ca::t_float32: 
      return inner_histo5<float>(input, output, min, max, nbins, accumulate);
    case ca::t_float64: 
      return inner_histo5<double>(input, output, min, max, nbins, accumulate);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

BOOST_PYTHON_FUNCTION_OVERLOADS(histo5_overloads, histo5, 5, 6)

template <typename T>
static object inner_histo3a (tp::const_ndarray input, object max) {
  T tmax = extract<T>(max);
  uint32_t size = (uint32_t)(tmax + 1);
  tp::ndarray out(ca::t_uint64, size);
  blitz::Array<uint64_t,1> out_ = out.bz<uint64_t,1>();
  ip::histogram(input.bz<T,2>(), out_, (T)0, tmax, size, false);
  return out.self();
}

static object histo3a (tp::const_ndarray input, object max) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_int8: return inner_histo3a<int8_t>(input, max);
    case ca::t_int16: return inner_histo3a<int16_t>(input, max);
    case ca::t_int32: return inner_histo3a<int32_t>(input, max);
    case ca::t_int64: return inner_histo3a<int64_t>(input, max);
    case ca::t_uint8: return inner_histo3a<uint8_t>(input, max);
    case ca::t_uint16: return inner_histo3a<uint16_t>(input, max);
    case ca::t_uint32: return inner_histo3a<uint32_t>(input, max);
    case ca::t_uint64: return inner_histo3a<uint64_t>(input, max);
    case ca::t_float32: return inner_histo3a<float>(input, max);
    case ca::t_float64: return inner_histo3a<double>(input, max);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

template <typename T>
static object inner_histo4a (tp::const_ndarray input, object min, object max) {
  T tmin = extract<T>(min);
  T tmax = extract<T>(max);
  int64_t size = (int64_t)(tmax - tmin + 1);
  tp::ndarray out(ca::t_uint64, size);
  blitz::Array<uint64_t,1> out_ = out.bz<uint64_t,1>();
  ip::histogram(input.bz<T,2>(), out_, tmin, tmax, size, false);
  return out.self();
}

static object histo4a (tp::const_ndarray input, object min, object max) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_int8: 
      return inner_histo4a<int8_t>(input, min, max);
    case ca::t_int16: 
      return inner_histo4a<int16_t>(input, min, max);
    case ca::t_int32: 
      return inner_histo4a<int32_t>(input, min, max);
    case ca::t_int64: 
      return inner_histo4a<int64_t>(input, min, max);
    case ca::t_uint8: 
      return inner_histo4a<uint8_t>(input, min, max);
    case ca::t_uint16: 
      return inner_histo4a<uint16_t>(input, min, max);
    case ca::t_uint32: 
      return inner_histo4a<uint32_t>(input, min, max);
    case ca::t_uint64: 
      return inner_histo4a<uint64_t>(input, min, max);
    case ca::t_float32: 
      return inner_histo4a<float>(input, min, max);
    case ca::t_float64: 
      return inner_histo4a<double>(input, min, max);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

template <typename T>
static object inner_histo5a (tp::const_ndarray input, object min, object max,
    uint32_t nbins) {
  T tmin = extract<T>(min);
  T tmax = extract<T>(max);
  tp::ndarray out(ca::t_uint64, nbins);
  blitz::Array<uint64_t,1> out_ = out.bz<uint64_t,1>();
  ip::histogram(input.bz<T,2>(), out_, tmin, tmax, nbins, false);
  return out.self();
}

static object histo5a (tp::const_ndarray input, 
    object min, object max, uint32_t nbins) {
  const ca::typeinfo& info = input.type();
  switch (info.dtype) {
    case ca::t_int8: 
      return inner_histo5a<int8_t>(input, min, max, nbins);
    case ca::t_int16: 
      return inner_histo5a<int16_t>(input, min, max, nbins);
    case ca::t_int32: 
      return inner_histo5a<int32_t>(input, min, max, nbins);
    case ca::t_int64: 
      return inner_histo5a<int64_t>(input, min, max, nbins);
    case ca::t_uint8: 
      return inner_histo5a<uint8_t>(input, min, max, nbins);
    case ca::t_uint16: 
      return inner_histo5a<uint16_t>(input, min, max, nbins);
    case ca::t_uint32: 
      return inner_histo5a<uint32_t>(input, min, max, nbins);
    case ca::t_uint64: 
      return inner_histo5a<uint64_t>(input, min, max, nbins);
    case ca::t_float32: 
      return inner_histo5a<float>(input, min, max, nbins);
    case ca::t_float64: 
      return inner_histo5a<double>(input, min, max, nbins);
    default:
      PYTHON_ERROR(TypeError, "unsupported histogram operation for type '%s'", info.str().c_str());
  }
}

void bind_ip_histogram() {
  def("histogram_", &histo2, histo2_overloads((arg("src"), arg("histo"), arg("accumulate")=false), "Compute an histogram of a 2D array. The histogram must have a size of 2^N-1 elements, where N is the number of bits in input. If the accumulate flag is set (defaults to False), then I accumulate instead of resetting the histogram."));
  
  def("histogram_", &histo3, histo3_overloads((arg("src"), arg("histo"), arg("max"), arg("accumulate")=false), "Compute an histogram of a 2D array.\nsrc elements are in range [0, max] (max >= 0)\nhisto must have a size of max elements"));
  
  def("histogram_", &histo4, histo4_overloads((arg("src"), arg("histo"), arg("min"), arg("max"), arg("accumulate")=false), "Compute an histogram of a 2D array.\nsrc elements are in range [min, max] (max >= min)\nhisto must have a size of max-min elements"));
  
  def("histogram_", &histo5, histo5_overloads((arg("src"), arg("histo"), arg("min"), arg("max"), arg("nb_bins"), arg("accumulate")=false), "Compute an histogram of a 2D array.\nsrc elements are in range [min, max] (max >= min)\nhisto must have a size of nb_bins elements")); 
  
  def("histogram", &histo1, (args("src")), "Compute an histogram of a 2D array");
  def("histogram", &histo3a, (arg("src"), arg("max")), "Return an histogram of a 2D array.\nsrc elements are in range [0, max] (max >= 0)\n");
  
  def("histogram", &histo4a, (arg("src"), arg("min"), arg("max")), "Return an histogram of a 2D array.\nsrc elements are in range [min, max] (max >= min)\n");
  
  def("histogram", &histo5a, (arg("src"), arg("min"), arg("max"), arg("nb_bins")), "Return an histogram of a 2D array.\nsrc elements are in range [min, max] (max >= min)\n");
}

File: python/bob/ip/test/histo.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 :
# Francois Moulin <Francois.Moulin@idiap.ch>
# Mon Apr 18 16:08:34 2011 +0200
#
# Copyright (C) 2011-2012 Idiap Research Institute, Martigny, Switzerland
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""Tests histogram computation 
"""

import os, sys
import unittest
import bob
import random
import numpy
import pkg_resources

def F(f):
  """Returns the test file on the "data" subdirectory"""
  return pkg_resources.resource_filename(__name__, os.path.join('data', f))

def load_gray(relative_filename):
  # Please note our PNG loader will always load in RGB, but since that is a
  # grayscaled version of the image, I just select one of the planes. 
  filename = os.path.join('histo', relative_filename)
  array = bob.io.load(F(filename))
  return array[0,:,:] 

def random_int(array, min_value, max_value):
  for i in range(0,array.shape()[0]):
    for j in range(0,array.shape()[1]):
      array[i, j] = random.randint(min_value, max_value)
        
def random_float(array, min_value, max_value):
  for i in range(0,array.shape()[0]):
    for j in range(0,array.shape()[1]):
      array[i, j] = random.uniform(min_value, max_value) 


class HistogramTest(unittest.TestCase):
  """Performs various histogram tests."""

  def test01_uint8_histoPython(self):
    """Compute the histogram of a uint8 image"""
    input_image = load_gray('image.ppm')
    

    histo1 = bob.ip.histogram(input_image)
    histo2 = bob.ip.histogram(input_image, 255)
    histo3 = bob.ip.histogram(input_image, 0, 255)
    histo4 = bob.ip.histogram(input_image, 0, 255, 256)
    
    histo5 = numpy.ndarray((256,), 'uint64')
    histo6 = numpy.ndarray((256,), 'uint64')
    histo7 = numpy.ndarray((256,), 'uint64')
    histo8 = numpy.ndarray((256,), 'uint64')
    
    bob.ip.histogram_(input_image, histo5)
    bob.ip.histogram_(input_image, histo6, 255)
    bob.ip.histogram_(input_image, histo7, 0, 255)
    bob.ip.histogram_(input_image, histo8, 0, 255, 256)
    
    # Save the computed data
    #bob.io.save(histo1, os.path.join('histo','image_histo.hdf5'))
    
    histo_ref = bob.io.load(F(os.path.join('histo','image_histo.hdf5')))

    self.assertTrue(input_image.size == histo1.sum())
    self.assertTrue(input_image.size == histo2.sum())
    self.assertTrue(input_image.size == histo3.sum())
    self.assertTrue(input_image.size == histo4.sum())
    self.assertTrue(input_image.size == histo5.sum())
    self.assertTrue(input_image.size == histo6.sum())
    self.assertTrue(input_image.size == histo7.sum())
    self.assertTrue(input_image.size == histo8.sum())
    self.assertTrue((histo_ref == histo1).all())
    self.assertTrue((histo_ref == histo2).all())
    self.assertTrue((histo_ref == histo3).all())
    self.assertTrue((histo_ref == histo4).all())
    self.assertTrue((histo_ref == histo5).all())
    self.assertTrue((histo_ref == histo6).all())
    self.assertTrue((histo_ref == histo7).all())
    self.assertTrue((histo_ref == histo8).all())
    
  def test02_uint16_histoPython(self):
    """Compute the histogram of a uint16 random array"""
    
    # Generate random uint16 array
    #input_array = numpy.ndarray((50, 70), 'uint16')
    #random_int(input_array, 0, 65535)
    #bob.io.save(input_array, os.path.join('histo','input_uint16.hdf5'))
    
    input_array = bob.io.load(F(os.path.join('histo','input_uint16.hdf5')))
    
    histo1 = bob.ip.histogram(input_array)
    histo2 = bob.ip.histogram(input_array, 65535)
    histo3 = bob.ip.histogram(input_array, 0, 65535)
    histo4 = bob.ip.histogram(input_array, 0, 65535, 65536)
    
    histo5 = numpy.ndarray((65536,), 'uint64')
    histo6 = numpy.ndarray((65536,), 'uint64')
    histo7 = numpy.ndarray((65536,), 'uint64')
    histo8 = numpy.ndarray((65536,), 'uint64')
    
    bob.ip.histogram_(input_array, histo5)
    bob.ip.histogram_(input_array, histo6, 65535)
    bob.ip.histogram_(input_array, histo7, 0, 65535)
    bob.ip.histogram_(input_array, histo8, 0, 65535, 65536)
    
    # Save computed data
    #bob.io.save(histo1, os.path.join('histo','input_uint16.histo.hdf5'))
    
    histo_ref = bob.io.load(F(os.path.join('histo','input_uint16.histo.hdf5')))
    
    self.assertTrue(input_array.size == histo1.sum())
    self.assertTrue(input_array.size == histo2.sum())
    self.assertTrue(input_array.size == histo3.sum())
    self.assertTrue(input_array.size == histo4.sum())
    self.assertTrue(input_array.size == histo5.sum())
    self.assertTrue(input_array.size == histo6.sum())
    self.assertTrue(input_array.size == histo7.sum())
    self.assertTrue(input_array.size == histo8.sum())
    self.assertTrue((histo_ref == histo1).all())
    self.assertTrue((histo_ref == histo2).all())
    self.assertTrue((histo_ref == histo3).all())
    self.assertTrue((histo_ref == histo4).all())
    self.assertTrue((histo_ref == histo5).all())
    self.assertTrue((histo_ref == histo6).all())
    self.assertTrue((histo_ref == histo7).all())
    self.assertTrue((histo_ref == histo8).all())
    
  def test03_float_histoPython(self):
    """Compute the histogram of a float random array"""
    
    # Generate random float32 array
    #input_array = numpy.ndarray((50, 70), 'float32')
    #random_float(input_array, 0, 1)
    #bob.io.save(input_array, os.path.join('histo','input_float.hdf5'))
    
    input_array = bob.io.load(F(os.path.join('histo','input_float.hdf5')))
    
    histo2 = numpy.ndarray((10,), 'uint64')
    
    histo1 = bob.ip.histogram(input_array, 0, 1, 10)
    bob.ip.histogram_(input_array, histo2, 0, 1, 10)
    
    # Save computed data
    #bob.io.save(histo1,os.path.join('histo','input_float.histo.hdf5'))
    
    histo_ref = bob.io.load(F(os.path.join('histo','input_float.histo.hdf5')))
    
    self.assertTrue(input_array.size == histo1.sum())
    self.assertTrue(input_array.size == histo2.sum())
    self.assertTrue((histo_ref == histo1).all())
    self.assertTrue((histo_ref == histo2).all())  

  def test04_int32_histoPython(self):
    """Compute the histogram of a int32 random array"""
    
    # Generate random int32 array
    #input_array = numpy.ndarray((50, 70), 'int32')
    #random_int(input_array, -20,20)
    #bob.io.save(input_array,os.path.join('histo','input_int32.hdf5'))
    
    input_array = bob.io.load(F(os.path.join('histo','input_int32.hdf5')))
    
    histo2 = numpy.ndarray((41,), 'uint64')

    histo1 = bob.ip.histogram(input_array, -20, 20, 41)
    bob.ip.histogram_(input_array, histo2, -20, 20, 41)
    
    # Save computed data
    #bob.io.save(histo, os.path.join('histo','input_int32.histo.hdf5'))
    
    histo_ref = bob.io.load(F(os.path.join('histo','input_int32.histo.hdf5')))
    
    self.assertTrue(input_array.size == histo1.sum())
    self.assertTrue(input_array.size == histo2.sum())
    self.assertTrue((histo_ref == histo1).all())
    self.assertTrue((histo_ref == histo2).all())
    
  def test05_uint32_accumulate_histoPython(self):
    """Accumulate the histogram of a int32 random array"""
    
    input_array = bob.io.load(F(os.path.join('histo','input_int32.hdf5')))
    
    histo = bob.ip.histogram(input_array, -20, 20, 41)
    
    bob.ip.histogram_(input_array, histo, -20, 20, 41, True)

    histo_ref = bob.io.load(F(os.path.join('histo','input_int32.histo.hdf5')))

    self.assertTrue(input_array.size * 2 == histo.sum())
    self.assertTrue((histo_ref * 2 == histo).all())

  def test06_uint16(self):
    """Simple test as described in ticket #101"""
    x = numpy.array([[-1., 1.],[-1., 1.]])
    res = bob.ip.histogram(x, -2, +2, 2)

    histo_ref = numpy.array([2, 2], 'uint64')
    self.assertTrue((histo_ref == res).all())
